# TrackPlotR

When analyzing Epigenome sequencing data, a common task is to visualize the signal intensity detected at different position of the genome, the distribution and structure of various cis-elements/genes in the corresponding genome region, and flexibly combine them together. [UCSC Genome Browser](https://genome.ucsc.edu) and [WashU Epigenome Browser](http://epigenomegateway.wustl.edu/browser) are the greatest and most famous online tools that help to do the visualization task. However, exporting track plots generated by these online tools in a lossless and personalized manner can often be exhausting. Based on ggplot2, a powerful data visualization toolkit in R language, TrackPlotR is designed to generate attractive and high-quality genomic track plots using the most commonly employed data formats in a concise, elegant and highly customizable manner.

## Installation

TrackPlotR can be installed from GitHub:

```{R}
devtools::install_github('yimingsun12138/TrackPlotR')
```

## Requirement

TrackPlotR was designed with the aim of retaining high scalability while minimizing dependency on other packages as much as possible. The packages listed below are necessary for the installation of TrackPlotR:

- [ggplot2](https://cran.r-project.org/web/packages/ggplot2/index.html)

- [tidyr](https://cran.r-project.org/web/packages/tidyr/index.html)

- [rtracklayer](https://bioconductor.org/packages/release/bioc/html/rtracklayer.html)

## Document

TrackPlotR currently provides functions for visualizing signal coverage, genome features and transcripts. Meanwhile, TrackPlotR also provides build-in data to help get started quickly.

```{R}
library(ggplot2)
library(rtracklayer)
library(TrackPlotR)
library(patchwork)

#build-in example data
example_path <- system.file('extdata',package = 'TrackPlotR')
list.files(example_path)
```

```
## [1] "Homo_sapiens.GRCh38.105.subset.gtf" "S1_ATAC_coverage.bw"               
## [3] "S1_ATAC_peak.bed"                   "S2_ATAC_coverage.bw"               
## [5] "S2_ATAC_peak.bed"                   "S3_ATAC_coverage.bw"               
## [7] "S3_ATAC_peak.bed"                   "S4_ATAC_coverage.bw"
```

The build-in data includes Tn5 insertion site coverage for 4 samples (ATAC_coverage.bw), ATAC-seq peak region for 3 samples, and a human gene annotation file (Ensembl release 105).

### Coverage track plot

First load the BigWig file into R using `load_BigWig` and assign a sample name for each file:

```{R}
coverage_info <- paste(example_path,c('S1_ATAC_coverage.bw','S2_ATAC_coverage.bw','S3_ATAC_coverage.bw','S4_ATAC_coverage.bw'),sep = '/')
coverage_info <- load_BigWig(file_path = coverage_info,file_name = c('S1','S2','S3','S4'))
```

Plot the coverage signal in the specified genome region:

```{R}
region <- as('chr3:58510000-58690000','GRanges')
p1 <- coverage_vis_region(coverage_table = coverage_info,region = region,y_lim = 0.03,sample_order = c('S1','S2','S3','S4'),col_pal = c("#D51F26","#272E6A","#208A42","#89288F"))
print(p1 + theme(aspect.ratio = 0.1))
```

![coverage plot region](https://user-images.githubusercontent.com/32930896/252123733-1ab14b38-92d4-4773-8e63-f7a4b045ac1d.png "coverage plot region")

Sometimes we are interested in the chromatin accessibility landscape around a certain gene, and TrackPlotR provides a quick method for the visualization task:

```{R}
human_anno <- paste(example_path,'Homo_sapiens.GRCh38.105.subset.gtf',sep = '/')
human_anno <- rtracklayer::import(con = human_anno,format = 'gtf')
p1 <- coverage_vis_gene(coverage_table = coverage_info,gene_anno = human_anno,column_name = 'gene_name',gene = 'FAM107A',up_extend = 50000,down_extend = 50000,y_lim = 0.03,sample_order = c('S1','S2','S3','S4'),col_pal = c("#D51F26","#272E6A","#208A42","#89288F"))
print(p1 + theme(aspect.ratio = 0.1))
```

![coverage plot gene](https://user-images.githubusercontent.com/32930896/252123758-c10e9ddd-6b5a-472a-9956-a064f3304e3f.png "coverage plot gene")

### Feature track plot

TrackPlotR also provides methods for visualizing cis-elements (features) distribution in the specified genome region or around a certain gene. These cis-elements can be peak list, enhancers, human accelerated regions (HARs) and so on, and should be provided in the bed or GRanges/GRangesList format.

```{R}
#load peak files and assign sample name
peak_info <- paste(example_path,c('S1_ATAC_peak.bed','S2_ATAC_peak.bed','S3_ATAC_peak.bed'),sep = '/')
names(peak_info) <- c('S1','S2','S3')
```

```{R}
#feature plot in the specified genome region
p2 <- feature_vis_region(features = peak_info,region = region,col_pal = c("#D51F26","#272E6A","#208A42"))
print(p2 + theme(aspect.ratio = 0.2))
```

![feature plot region](https://user-images.githubusercontent.com/32930896/252123773-bd27f960-b0a2-4597-a177-10f8101c11b5.png "feature plot region")

```{R}
#feature plot around a certain gene
p2 <- feature_vis_gene(features = peak_info,gene_anno = human_anno,column_name = 'gene_name',gene = 'FAM107A',up_extend = 50000,down_extend = 50000,col_pal = c("#D51F26","#272E6A","#208A42"))
print(p2 + theme(aspect.ratio = 0.2))
```

![feature plot gene](https://user-images.githubusercontent.com/32930896/252123795-28cd03a0-4a02-450d-b092-f2b71fe4fdcd.png "feature plot gene")

### Transcript track plot

TrackPlotR also provides methods for visualizing transcripts/genes structure and distribution in the specified genome region or around a certain gene. Gene annotation file can be provided in the GTF or GRanges format and must contain 3 columns: type, gene_id and transcript_id.

```{R}
#gene plot in the specified genome region
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'gene',show_name = 'gene_id')
print(p3 + theme(aspect.ratio = 0.2))
```

![transcript plot region](https://user-images.githubusercontent.com/32930896/252123811-8cadc557-200b-4185-afc2-7f5f2c89461e.png "transcript plot region")

```{R}
#plot a certain gene only
p3 <- gene_track_vis(gene_anno = human_anno,column_name = 'gene_name',gene = 'FAM107A',up_extend = 50000,down_extend = 50000,show_only = TRUE)
print(p3 + theme(aspect.ratio = 0.2))
```

![transcript plot gene](https://user-images.githubusercontent.com/32930896/252123836-1bf6b9af-5429-4726-8af8-5268d662b334.png "transcript plot gene")

In addition to be displayed by gene, the gene annotation file can also be displayed on a transcript basis.

```{R}
#display by transcript
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'transcript',show_name = 'transcript_name')
print(p3 + theme(aspect.ratio = 0.5))
```

![transcript plot by transcript](https://user-images.githubusercontent.com/32930896/252123864-af00b233-53c4-46c2-b6f1-f473b379d240.png "transcript plot by transcript")

### Plotting style

Usually, we would like to observe different track plots together, so it is particularly important to set a uniform style for different track plots. Currently, TrackPlotR has only one build-in plotting style, and more styles will be added in the future.

```{R}
p1 <- coverage_vis_region(coverage_table = coverage_info,region = region,y_lim = 0.03,sample_order = c('S1','S2','S3','S4'),col_pal = c("#D51F26","#272E6A","#208A42","#89288F"),style = 'style_1')
p2 <- feature_vis_region(features = peak_info,region = region,col_pal = c("#D51F26","#272E6A","#208A42"),style = 'style_1')
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'gene',show_name = 'gene_name',style = 'style_1')
```

Thanks to the high flexibility and scalability of ggplot2, we can make more personalized modifications to the images returned by TrackPlotR and stitch them together easily.

```{R}
p1 <- p1 + theme(axis.ticks.x = element_blank(),axis.text.x = element_blank(),axis.title.x = element_blank())
p2 <- p2 + theme(axis.ticks.x = element_blank(),axis.text.x = element_blank(),axis.title.x = element_blank(),legend.position = 'none')
print(p1 + p2 + p3 + plot_layout(ncol = 1,heights = c(0.5,0.1,0.15)))
```

![style_1 plot](https://user-images.githubusercontent.com/32930896/252123890-43537c55-7e6c-4120-96b8-3b4dd1396a6f.png "style_1 plot")

### Display mode

Feature track plot and transcript track plot can be displayed with different row compression densities.

```{R}
# full transcripts
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'gene',show_name = 'gene_name',display_mode = 'full')
print(p3 + theme(aspect.ratio = 0.3))
```

![full transcript plot](https://user-images.githubusercontent.com/32930896/252123906-2dcf7b42-77d6-40c4-a019-085650fedc0e.png "full transcript plot")

In the full mode, each transcript/gene will occupy a single row.

```{R}
# squish transcripts
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'gene',show_name = 'gene_name',display_mode = 'squish')
print(p3 + theme(aspect.ratio = 0.2))
```

![squish transcript plot](https://user-images.githubusercontent.com/32930896/252123917-6b7b6006-c548-4058-8588-197890db3930.png "squish transcript plot")

In the squish mode, transcripts/genes that do not overlap with each other can occupy the same row.

```{R}
# collapse transcripts
p3 <- transcript_vis_region(gene_anno = human_anno,region = region,display_by = 'gene',show_name = NULL,display_mode = 'collapse')
print(p3 + theme(aspect.ratio = 0.1))
```

![collapse transcript plot](https://user-images.githubusercontent.com/32930896/252123932-b271be59-b116-4b5d-83e6-6a524ee54e2f.png "collapse transcript plot")

```{R}
# collapse features
p2 <- feature_vis_region(features = peak_info,region = region,col_pal = c("#D51F26","#272E6A","#208A42"),collapse_features = TRUE,overlap_col = 'darkgrey')
print(p2 + theme(aspect.ratio = 0.1))
```

![collapse feature plot](https://user-images.githubusercontent.com/32930896/252123944-faccb7ad-5fce-4811-88f9-4b8e104ad241.png "collapse feature plot")

In the collapse mode, all transcripts/genes and features will be plotted in the same row.

## Planning

- Link track visualization
- Highlight certain regions on the track plot
- Pile-Up plot using sequencing fragments
- Hi-C support