#' Coverage track visualization in the specified genome region
#' 
#' @description
#' Generate coverage track plot in the specified genome region.
#' 
#' @param coverage_table Genome coverage table generated by `load_BigWig`.
#' @param region Genome region used to generate the coverage track plot, must be a GRanges object.
#' @param up_extend How many base pairs to extend upstream the region?
#' @param down_extend How many base pairs to extend downstream the region?
#' @param style Plot style provided by package, check the [document](https://github.com/yimingsun12138/TrackPlotR) for details.
#' @param y_lim The numerical y-axis limit for coverage track plot.
#' @param sample_order The sample order when generating the coverage track.
#' @param col_pal A custom palette used to override coloring for samples.
#' 
#' @return A ggplot object.
#' 
#' @export
coverage_vis_region <- function(coverage_table,
                                region,
                                up_extend = 0,
                                down_extend = 0,
                                style = c('style_1'),
                                y_lim = NULL,
                                sample_order = NULL,
                                col_pal = NULL){
  
  #check parameter
  if(base::class(coverage_table) != 'data.frame'){
    base::stop('coverage_table must be a data.frame object!')
  }
  
  if(!base::all(c('seqnames','start','end','score','Sample') %in% base::colnames(coverage_table))){
    base::stop('missing columns in coverage_table!')
  }
  
  if(base::class(region) != 'GRanges'){
    base::stop('region must be a GRanges object!')
  }else{
    if(base::length(region) != 1){
      base::stop('only 1 region required!')
    }
  }
  
  #extend region
  IRanges::start(region) <- IRanges::start(region) - up_extend
  IRanges::end(region) <- IRanges::end(region) + down_extend
  
  #check style
  style <- style[1]
  if(!(style %in% c('style_1'))){
    base::stop('invalid style!')
  }
  
  #basic plot
  coverage_plot <- coverage_vis_basic(coverage_table = coverage_table,
                                      region = region,
                                      y_lim = y_lim,
                                      sample_order = sample_order,
                                      col_pal = col_pal)
  
  #style
  if(style == 'style_1'){
    coverage_plot <- coverage_plot
  }
  
  #return
  return(coverage_plot)
}

#' Coverage track visualization around specified gene region
#' 
#' @description
#' Generate coverage track plot around specified gene region.
#' 
#' @param coverage_table Genome coverage table generated by `load_BigWig`.
#' @param gene_anno Gene annotation (GTF) file path or a GRanges object which stores the gene annotation information. Ensembl GTF file is recommended.
#' @param column_name Which column in gene_anno to be searched for the specified gene name?
#' @param gene The name of the specified gene.
#' @param up_extend How many base pairs to extend upstream the gene?
#' @param down_extend How many base pairs to extend downstream the gene?
#' @param style Plot style provided by package, check the [document](https://github.com/yimingsun12138/TrackPlotR) for details.
#' @param y_lim The numerical y-axis limit for coverage track plot.
#' @param sample_order The sample order when generating the coverage track.
#' @param col_pal A custom palette used to override coloring for samples.
#' 
#' @return A ggplot object.
#' 
#' @export
coverage_vis_gene <- function(coverage_table,
                              gene_anno,
                              column_name,
                              gene,
                              up_extend = 3000,
                              down_extend = 3000,
                              style = c('style_1'),
                              y_lim = NULL,
                              sample_order = NULL,
                              col_pal = NULL){
  
  #check parameter
  if(base::class(gene_anno) != 'GRanges'){
    gene_anno <- rtracklayer::import(con = gene_anno,format = 'gtf')
  }
  
  if(!base::all(c('type',column_name) %in% base::colnames(gene_anno@elementMetadata))){
    base::stop(base::paste0('gene_anno must contain two columns: type and ',column_name,'!'))
  }
  
  #get gene region
  idx <- base::which((gene_anno$type == 'gene') & (gene_anno@elementMetadata[,column_name] == gene))
  if(base::length(idx) == 0){
    base::stop(base::paste(gene,'not found in the annotation!',sep = ' '))
  }
  if(base::length(idx) > 1){
    base::stop('multiple gene regions detected in the annotation!')
  }
  
  region <- gene_anno[idx]
  if(region@strand == '-'){
    IRanges::start(region) <- IRanges::start(region) - down_extend
    IRanges::end(region) <- IRanges::end(region) + up_extend
  }else{
    IRanges::start(region) <- IRanges::start(region) - up_extend
    IRanges::end(region) <- IRanges::end(region) + down_extend
  }
  
  #plot
  coverage_plot <- coverage_vis_region(coverage_table = coverage_table,
                                       region = region,
                                       up_extend = 0,
                                       down_extend = 0,
                                       style = style,
                                       y_lim = y_lim,
                                       sample_order = sample_order,
                                       col_pal = col_pal)
  
  #return
  return(coverage_plot)
}

#' Feature track visualization in the specified genome region
#' 
#' @description
#' Generate feature track plot in the specified genome region.
#' 
#' @param features A list of bed file paths or a GRanges/GRangesList object which stores the coordinates of genome features.
#' @param region Genome region used to generate the feature track plot, must be a GRanges object.
#' @param up_extend How many base pairs to extend upstream the region?
#' @param down_extend How many base pairs to extend downstream the region?
#' @param collapse_features Whether collapse the feature track, default is FALSE.
#' @param style Plot style provided by package, check the [document](https://github.com/yimingsun12138/TrackPlotR) for details.
#' @param col_pal A custom palette used to override coloring for features.
#' @param overlap_col Color for overlapped region of different features if collapse_features is set to TRUE.
#' @param segment_size The thickness (linewidth) of the feature segment.
#' 
#' @return A ggplot object.
#' 
#' @export
feature_vis_region <- function(features,
                               region,
                               up_extend = 0,
                               down_extend = 0,
                               collapse_features = FALSE,
                               style = c('style_1'),
                               col_pal = NULL,
                               overlap_col = 'darkgrey',
                               segment_size = 2){
  
  #check parameter
  if(base::grepl(pattern = 'GRanges',x = base::class(features),fixed = TRUE)){
    features <- features
  }else{
    features <- base::unlist(features)
    if(base::class(features) == 'character'){
      feature_name <- base::names(features)
      features <- base::do.call(what = GenomicRanges::GRangesList,args = base::lapply(X = features,FUN = rtracklayer::import.bed))
      base::names(features) <- feature_name
    }else{
      base::stop('features must be a list of bed file paths or a GRanges/GRangesList object!')
    }
  }
  
  if(base::class(region) != 'GRanges'){
    base::stop('region must be a GRanges object!')
  }else{
    if(base::length(region) != 1){
      base::stop('only 1 region required!')
    }
  }
  
  #extend region
  IRanges::start(region) <- IRanges::start(region) - up_extend
  IRanges::end(region) <- IRanges::end(region) + down_extend
  
  #check style
  style <- style[1]
  if(!(style %in% c('style_1'))){
    base::stop('invalid style!')
  }
  
  #basic plot
  feature_plot <- feature_vis_basic(Ranges = features,
                                    region = region,
                                    collapse_range = collapse_features,
                                    col_pal = col_pal,
                                    overlap_col = overlap_col,
                                    segment_size = segment_size)
  
  #style
  if(style == 'style_1'){
    feature_plot <- feature_plot
  }
  
  #return
  return(feature_plot)
}

#' Feature track visualization around specified gene region
#' 
#' @description
#' Generate feature track plot around specified gene region.
#' 
#' @param features A list of bed file paths or a GRanges/GRangesList object which stores the coordinates of genome features.
#' @param gene_anno Gene annotation (GTF) file path or a GRanges object which stores the gene annotation information. Ensembl GTF file is recommended.
#' @param column_name Which column in gene_anno to be searched for the specified gene name?
#' @param gene The name of the specified gene.
#' @param up_extend How many base pairs to extend upstream the gene?
#' @param down_extend How many base pairs to extend downstream the gene?
#' @param collapse_features Whether collapse the feature track, default is FALSE.
#' @param style Plot style provided by package, check the [document](https://github.com/yimingsun12138/TrackPlotR) for details.
#' @param col_pal A custom palette used to override coloring for features.
#' @param overlap_col Color for overlapped region of different features if collapse_features is set to TRUE.
#' @param segment_size The thickness (linewidth) of the feature segment.
#' 
#' @return A ggplot object.
#' 
#' @export
feature_vis_gene <- function(features,
                             gene_anno,
                             column_name,
                             gene,
                             up_extend = 3000,
                             down_extend = 3000,
                             collapse_features = FALSE,
                             style = c('style_1'),
                             col_pal = NULL,
                             overlap_col = 'darkgrey',
                             segment_size = 2){
  
  #check parameter
  if(base::class(gene_anno) != 'GRanges'){
    gene_anno <- rtracklayer::import(con = gene_anno,format = 'gtf')
  }
  
  if(!base::all(c('type',column_name) %in% base::colnames(gene_anno@elementMetadata))){
    base::stop(base::paste0('gene_anno must contain two columns: type and ',column_name,'!'))
  }
  
  #get gene region
  idx <- base::which((gene_anno$type == 'gene') & (gene_anno@elementMetadata[,column_name] == gene))
  if(base::length(idx) == 0){
    base::stop(base::paste(gene,'not found in the annotation!',sep = ' '))
  }
  if(base::length(idx) > 1){
    base::stop('multiple gene regions detected in the annotation!')
  }
  
  region <- gene_anno[idx]
  if(region@strand == '-'){
    IRanges::start(region) <- IRanges::start(region) - down_extend
    IRanges::end(region) <- IRanges::end(region) + up_extend
  }else{
    IRanges::start(region) <- IRanges::start(region) - up_extend
    IRanges::end(region) <- IRanges::end(region) + down_extend
  }
  
  #plot
  feature_plot <- feature_vis_region(features = features,
                                     region = region,
                                     up_extend = 0,
                                     down_extend = 0,
                                     collapse_features = collapse_features,
                                     style = style,
                                     col_pal = col_pal,
                                     overlap_col = overlap_col,
                                     segment_size = segment_size)
  
  #return
  return(feature_plot)
}
